ffpyplayer/player/core.pyx

```diff
@@ -2250,6 +2250,100 @@ cdef class VideoState(object):
             self.request_thread_s(b'read:exit', b'Done')
         return 0
 
+    cdef int stream_select_program(VideoState self,
+                                   int requested_program) nogil except 1:
+        cdef unsigned int i
+        cdef unsigned int j
+        cdef AVProgram *p, *selected_program = NULL
+        cdef AVStream *st
+        cdef unsigned int nb_streams
+        cdef unsigned int stream_index
+        cdef int video_stream_index = -1
+        cdef int audio_stream_index = -1
+        cdef int subtitle_stream_index = -1
+        cdef int program = -1
+        cdef unsigned int nb_programs = self.ic.nb_programs
+
+        i = 0
+        while i < nb_programs:
+            p = self.ic.programs[i]
+
+            if p.id == requested_program:
+                selected_program = p
+                break
+
+            i += 1
+
+        if selected_program == NULL:
+            return -1
+
+        nb_streams = selected_program.nb_stream_indexes
+
+        j = 0
+        while j < nb_streams:
+            stream_index = selected_program.stream_index[j]
+            st = self.ic.streams[stream_index]
+
+            if st.codecpar.codec_type == AVMEDIA_TYPE_VIDEO:
+                if video_stream_index == -1:
+                    video_stream_index = <int>stream_index
+            elif st.codecpar.codec_type == AVMEDIA_TYPE_AUDIO:
+                if audio_stream_index == -1:
+                    audio_stream_index = <int>stream_index
+            elif st.codecpar.codec_type == AVMEDIA_TYPE_SUBTITLE:
+                if subtitle_stream_index == -1:
+                    subtitle_stream_index = <int>stream_index
+
+            j += 1
+
+        if video_stream_index != -1:
+            self.stream_component_close(self.video_stream)
+            self.stream_component_open(video_stream_index)
+
+        if audio_stream_index != -1:
+            self.stream_component_close(self.audio_stream)
+            self.stream_component_open(audio_stream_index)
+
+        if subtitle_stream_index != -1:
+            self.stream_component_close(self.subtitle_stream)
+            self.stream_component_open(subtitle_stream_index)
+
+        return 0
+
+    cdef int stream_select_channel(VideoState self, int codec_type,
```

Comment:
Why not use `stream_cycle_channel` like in `request_channel`. As far as I can tell they are mostly the same?

Reply:
While stream_cycle_channel works ok when cycling from current to next, it doesn't work well when you need to select an arbitrary stream (/w requested_stream argument. I ran across this issue while shuffling between HLS streams and it seemed simpler to write separate functions for (arbitrary) stream and program selection instead of trying to jerry-rig stream_cycle_channel.

Reply:
Ok. Please also add that info to the docs because people will wonder why there are two functions that do similar things.

Otherwise, is the PR done?

Reply:
I would suggest making 'requested_stream' argument of stream_cycle_channel inert, since it is essentially broken anyway. If you agree I can add the changes to the PR and update the docs.

Reply:
Yeah, that sounds good. Although maybe just add a deprecation warning pointing to use the new method when `requested_stream` is not `-1`, to not break backward compat and we can remove it altogether in a later version.

Reply:
I removed `requested_stream` and modified `request_channel` so it calls `stream_select_channel` if `action='open'`. Should be fully backward compatible now.